### ISSUE TITLE
BUG: GSL callbacks cast params to the wrong struct with wrong entry levels, inert stop-loss exits, and undefined delete

### ISSUE THREAD

**Issue by jasonharris438** [jasonharris438/stochastic-models#41]
### **Description**

The trading-level code hands parameter structs to GSL callbacks through `void*`. Several callbacks cast that pointer to a different struct type. The reference members line up by position, so every cast compiles and runs. The callbacks then read the wrong parameters and return plausible, wrong numbers on the library's main output surface.

Each case reads an object through an unrelated type, which is undefined behavior. On common ABIs the wrong values are deterministic.

| Item | Defect | Effect |
|---|---|---|
| 1 | Entry-level callbacks read `b_star` as `r` | Every entry level uses a corrupted discount rate |
| 2 | Stop-loss exits dispatch to the no-stop-loss callback | The stop-loss model never runs |
| 3 | `OptimalTrading` is deleted with no virtual destructor | Undefined behavior on every entry and exit call |

### Item 1: Entry levels read `b_star` as the discount rate

The 4-double `instantaneousDifferential` overload builds `EntryLevelParams{optimizer, kernel, b_star, r, c}` and dispatches to `funcIntegrateF` / `funcIntegrateG` (`src/optimal_mean_reversion.cpp:34-43`). Those callbacks cast the pointer to `ExitLevelParams{optimizer, kernel, r, c}`. The slots shift by one: slot 3 reads `b_star` as `r`, and slot 4 reads `r` as `c`. F and G ignore `c`, so the effective corruption is the discount rate.

Five call sites pass `b_star` into that overload:

| Method | Site | Callback |
|---|---|---|
| `OptimalMeanReversion::d` (stop-loss) | `src/optimal_mean_reversion.cpp:184-186` | `funcG` |
| `OptimalMeanReversion::d` | `src/optimal_mean_reversion.cpp:210-212` | `funcG` |
| `OptimalMeanReversion::a` (stop-loss) | `src/optimal_mean_reversion.cpp:246-248` | `funcF` |
| `ExponentialMeanReversion::d` | `src/exponential_mean_reversion.cpp:150-152` | `funcG` |
| `ExponentialMeanReversion::a` | `src/exponential_mean_reversion.cpp:232-234` | `funcF` |

The same overload also serves `valueFunc`, which casts `EntryLevelParams` and needs `b_star`. That pairing is correct and must stay.

Affected public surface: `optimalEntry` and `optimalEntryLower` on both classes, all overloads. That is every entry level the library produces.

### Item 2: Stop-loss exits solve the wrong equation

Both `optimalExit(stop_loss, r, c)` methods allocate `ExitLevelStopLossParams{optimizer, kernel, stop_loss, r, c}` and pass `funcOptimalMeanReversionB` (`src/trading_levels.cpp:59-62`, `src/trading_levels_exponential.cpp:64-67`). That callback casts to `ExitLevelParams` and calls the no-stop-loss `b()` with `r = stop_loss` and `c = r`. The stop-loss branch never runs.

The correct callback exists and is dead: `funcOptimalMeanReversionStopLossB` (`src/trading_levels_params.cpp:44-48`). Nothing references it.

Affected public surface: `optimalExitLevel(mu, alpha, sigma, stop_loss, r, c)` and the stop-loss `optimalExit` methods.

The test hides the defect. `optimalExitLevelStopLossOutputTest` sets `stop_loss = 0.05` and `r = 0.05`, so the swapped slots read the same value. Its expected `0.46912012` is the no-stop-loss root with the scrambled cost `c = 0.05`, and it breaks the theory bound `b*_L <= b*` (`0.46912 > 0.466836`). The entry tests feed `b_star = 0.450895`, which is the correct stop-loss exit.

The stop-loss overload of `ExponentialMeanReversion::b` throws `std::logic_error` by design. So the exponential stop-loss exit must throw at entry, like its sibling stubs, and not enter the solver. Its entrypoint definition (`src/entrypoint_optimal_trading_levels.cpp:58-80`) is an orphan: defined, and never declared in `optimal_trading_levels.h`. Delete it.

### Item 3: Polymorphic delete with no virtual destructor

`include/stochastic_models/trading/optimal_trading.h` declares 15 pure virtual functions and no destructor. All four params-struct destructors (`src/trading_levels_params.cpp:3, 9, 15, 21`) run `delete optimizer;` through `const OptimalTrading*`. The compiler reports every site:

```
warning: deleting object of abstract class type 'OptimalTrading' which has
non-virtual destructor will cause undefined behavior
[-Wdelete-non-virtual-dtor]
```

This is undefined behavior on every entry and exit call. It is harmless today only because both derived classes have no members. The build hides the warning with `-Wno-delete-non-virtual-dtor` (`src/CMakeLists.txt:80`). This issue removes that suppression.

The fix is one line: `virtual ~OptimalTrading() = default;`. The `delete hitting_time_kernel` in the same destructors is correct, because `HittingTimeOrnsteinUhlenbeck` is not polymorphic.

### Corrected values

The current tests are regression locks. A replica of the buggy computation reproduces every locked entry and stop-loss-exit value. An independent computation of the Leung & Li equations (quadrature at 1e-12, roots verified against the smooth-pasting identities, stable to about 1e-6) gives the correct values:

| Case | Locked (wrong) | Correct |
|---|---|---|
| OU exit `b*` | 0.466836 | 0.4668359 (unaffected) |
| OU entry `d*` (`b* = 0.466836`) | 0.116948 | **0.1156811** |
| OU stop-loss exit `b*_L` (`L = 0.05`) | 0.46912012 | **0.4508948** |
| OU stop-loss entry `d*` (`b* = 0.450895`) | 0.136755 | **0.1362685** |
| OU stop-loss entry `a*` (`b* = 0.450895`) | 0.118451 | **0.1184365** |
| XOU exit `b*` | 1.4093 | 1.4092995 (unaffected) |
| XOU entry `d*` (`b* = 1.4093`) | 1.24096 | **1.2368866** |
| XOU entry-lower `a*` | 1.16016 | **no root in the bracket — typed exception** |

The XOU `a*` lock is the solver bracket endpoint: `mu - 4*sqrt(sigma^2/(2*alpha)) = 1.1601633`. The `a`-equation has no root inside that bracket. Its root lies near −11 in log-price, and a wider `a*` bracket is out of scope here. `brentSolver` throws `RootNotBracketedError` on such a bracket, so the corrected test asserts that type.

The stop-loss rows above sit at `stop_loss = r`, so the stop-loss fixtures move to `stop_loss = 0.04`, derived the same way: `b*_L = 0.4551908`, `d* = 0.1309298` (with `b_star = 0.455191`), and `a* = 0.1076331` (with `d_star = 0.13093`).

Four more fixtures sit on the corrupted path:

- `methodDOutputTest`: correct value `-129.0961`.
- `methodDStopLossOutputTest`: move to `stop_loss = 0.1`, correct value `-61.8326`. At the current `-0.3` the stop-loss value equals the plain value to six decimals, so the test cannot tell the two paths apart.
- `methodAStopLossOutputTest`: correct value `-132.5528`.
- `methodBelowVStopLossOutputTest`: call the stop-loss `V` overload at `x = 0.5`, correct value `0.4693219`. The test calls the plain overload today, and at `x = 0.15` the stop-loss branch returns `x - c`.

### **Acceptance Criteria**

- Every GSL callback casts its `void*` to the struct type that its call sites construct. No cross-type read remains.
- `OptimalTrading` declares a virtual destructor. The build is clean with `-Wdelete-non-virtual-dtor` active, and the `src/CMakeLists.txt` suppression is removed.
- `optimalExit(stop_loss, r, c)` on the OU class solves the stop-loss exit equation. On a fixture with `stop_loss != r`, its result differs from the no-stop-loss result.
- The exponential stop-loss exit throws `std::logic_error`, and its orphan entrypoint definition is deleted.
- Tests assert the corrected values above with `EXPECT_NEAR`, at tolerance 1e-4 for levels and 1e-3 for raw equation values. No fixture uses `stop_loss == r`. The XOU `a*` tests assert `RootNotBracketedError`.
- A consistency test asserts `stop_loss < a* < d* < b*_L < b*` on the OU set at `stop_loss = 0.04`: `0.04 < 0.1076331 < 0.1309298 < 0.4551908 < 0.4668359`.
- Every trading callback opens with a debug assertion on a type tag in its params struct. UBSan cannot detect this mismatch class, because the structs alias cleanly and none is polymorphic.
- The suite passes under ASan and UBSan.

### **Technical Approach**

Item 1: the five call sites drop the `b_star` argument and dispatch to the 3-double `instantaneousDifferential` overload, which builds the `ExitLevelParams` that the callbacks cast. The overload bodies and the `valueFunc` pairing stay untouched. Group 2: wire `funcOptimalMeanReversionStopLossB` on the OU class. Group 3: the one-line destructor. Tripwire: an `enum class` tag as the last member of each params struct, with a default member initializer, so every brace-init compiles unchanged.

#12's RAII refactor removes the defect class structurally: a typed trampoline makes a wrong cast unwritable. Land this fix first, so the refactor verifies against correct outputs.
